DelegatingFilterProxy looks at "registered" WebApplicationContexts

rwinch opened the pull request:
Previously when users registered an additional Filter in the
DispatcherServlet the DelegatingFilterProxy would not find it unless
the context attribute was set. The value for which was difficult to
remember and added quite a bit of additional "if" statements to setting
up Spring Security successfully.

This change allows registering additional known WebApplicationContext
instances. The DispatcherServlet uses this mechanism to register its
context. With that change DelegatingFilterProxy searches for the first
registered WebApplicationContext that contains the bean name.

This change also sets up the ability for Spring Security to lookup
Beans defined in the child context within the Tag libs.

Issue: SPR-13191

sbrannen commented on spring-web/src/main/java/org/springframework/web/context/support/WebApplicationContextUtils.java:
```diff
@@ -125,6 +136,63 @@ public static WebApplicationContext getWebApplicationContext(ServletContext sc,
+	 * but may be empty.
+	 * @since 4.2
+	 */
+	public static List<WebApplicationContext> getRegisteredWebApplicationContext(ServletContext sc) {
```
This method should be named `getRegisteredWebApplicationContexts` (i.e. plural).

rwinch replied:
Good catch! This is fixed now.

sbrannen commented on spring-web/src/main/java/org/springframework/web/context/support/WebApplicationContextUtils.java:
```diff
@@ -125,6 +136,63 @@ public static WebApplicationContext getWebApplicationContext(ServletContext sc,
 	}
 
+	/**
+	 * Registers a custom {@link WebApplicationContext} for this web app.
```
Spring's Javadoc convention is to use an imperative statement for the list line -- in other words, "Register ..." instead of "Registers ...". If you don't follow that rule, you run the risk of getting Juergenized. :wink:

rwinch replied:
Being Juergenized via proxy...I like it! This is fixed now :)

sbrannen commented on spring-web/src/main/java/org/springframework/web/context/support/WebApplicationContextUtils.java:
```diff
@@ -125,6 +136,63 @@ public static WebApplicationContext getWebApplicationContext(ServletContext sc,
+
+		@SuppressWarnings("unchecked")
+		LinkedHashSet<String> attrNames = (LinkedHashSet<String>) sc.getAttribute(APPLICATION_CONTEXT_ATTRIBUTES_ATTRIBUTE);
+		if(attrNames == null) {
```
Spring requires a space between `if`, `for`, etc. and the following parenthesis.

rwinch replied:
Thanks! This should be fixed now. Is there an IntelliJ code formatter available?

sbrannen replied:
There's a [Code Style](https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style) page.

Commit:
DelegatingFilterProxy looks at "registered" WebApplicationContexts

Previously when users registered an additional Filter in the
DispatcherServlet the DelegatingFilterProxy would not find it unless
the context attribute was set. The value for which was difficult to
remember and added quite a bit of additional "if" statements to setting
up Spring Security successfully.

This change allows registering additional known WebApplicationContext
instances. The DispatcherServlet uses this mechanism to register its
context. With that change DelegatingFilterProxy searches for the first
registered WebApplicationContext that contains the bean name.

This change also sets up the ability for Spring Security to lookup
Beans defined in the child context within the Tag libs.

Issue: SPR-13191

sbrannen commented:
Very interesting proposal!

Commit:
Add support for JdbcTempalte.queryForObject(sql,<primitive>.class)

Previously using jdbcTemplate.queryForObject(sql, int.class) produced
a TypeMismatchDataAccessException stating it could not convert an Integer
to an int.

This update allows conversion to the respective primitives instead of just
the wrapper Objects.

Issue: SPR-13220

rwinch commented:
Closing since this was resolved in https://github.com/spring-projects/spring-framework/commit/1fcd465f2d281a432eee79adfe1d4d8dd7ec3630
